SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -785,10 +843,109 @@ void USpatialReceiver::ApplyComponentData(Worker_EntityId EntityId, Worker_Compo
 	if (!TargetObject.IsValid())
 	{
-		UE_LOG(LogSpatialReceiver, Warning, TEXT("EntityId %lld, ComponentId %d, Offset %d - Could not find target object with given offset for Actor %s!"), EntityId, Data.component_id, Offset, *Channel->GetActor()->GetName());
+		// If we can't find this subobject, it's a dynamically attached object. Create it now.
+		TargetObject = NewObject<UObject>(Channel->GetActor(), ClassInfoManager->GetClassByComponentId(Data.component_id));
+
+		Channel->GetActor()->OnSubobjectCreatedFromReplication(TargetObject.Get());
+
+		PackageMap->ResolveSubobject(TargetObject.Get(), FUnrealObjectRef(EntityId, Offset));
+
+		Channel->CreateSubObjects.Add(TargetObject.Get());
+	}
+
+	ApplyComponentData(TargetObject.Get(), Channel, Data);
+}
+
+void USpatialReceiver::HandleDynamicAddComponent(Worker_AddComponentOp& Op)
+{
+	// If we are delegated a component that exists, we will receive an AddComponentOp.
+	// This will be a duplicate of the AddComponentOp we've already recieved in the AddEntity payload
+	// So it can be safely ignored.
+	if (StaticComponentView->HasComponent(Op.entity_id, Op.data.component_id))
+	{
+		return;
+	}
+
+	uint32 Offset = 0;
+	bool bFoundOffset = ClassInfoManager->GetOffsetByComponentId(Op.data.component_id, Offset);
+	if (!bFoundOffset)
+	{
+		UE_LOG(LogSpatialReceiver, Warning, TEXT("EntityId %lld, ComponentId %d - Could not find offset for component id when receiving dynamic AddComponent."), Op.entity_id, Op.data.component_id);
 		return;
 	}
 
+	// Object already exists, we can apply data directly. 
+	if (UObject* Object = PackageMap->GetObjectFromUnrealObjectRef(FUnrealObjectRef(Op.entity_id, Offset)).Get())
+	{
+		ApplyComponentData(Object, NetDriver->GetActorChannelByEntityId(Op.entity_id), Op.data);
+		return;
+	}
+
+	// Otherwise this is a dynamically attached component. We need to make sure we have all related components before creation.
+	PendingDynamicSubobjectComponents.Add(MakeTuple(Op.entity_id, Op.data.component_id), PendingAddComponentWrapper(Op.entity_id, Op.data.component_id, MakeUnique<DynamicComponent>(Op.data)));
+
+	const FClassInfo& Info = ClassInfoManager->GetClassInfoByComponentId(Op.data.component_id);
+
+	bool bReadyToCreate = true;
+	ForAllSchemaComponentTypes([&](ESchemaComponentType Type)
+	{
+		Worker_ComponentId ComponentId = Info.SchemaComponents[Type];
+
+		if (ComponentId == SpatialConstants::INVALID_COMPONENT_ID)
+		{
+			return;
+		}
+
+		if (!PendingDynamicSubobjectComponents.Contains(MakeTuple(Op.entity_id, ComponentId )))
+		{
+			bReadyToCreate = false;
+		}
+	});
+
+	if (bReadyToCreate)
+	{
+		AttachDynamicSubobject(Op.entity_id, Info);
+	}
+}
+
+void USpatialReceiver::AttachDynamicSubobject(Worker_EntityId EntityId, const FClassInfo& Info)
+{
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId).Get());
+
+	if (Actor == nullptr)
+	{
```

Comment:
Should we warn here? I would imagine we couldn't receive add component after remove entity?

Reply:
Yeah I think you're right. Will warn.